### ISSUE TITLE
Weird sampling from multinomial_alias_draw

### ISSUE THREAD

**Issue by LeviViana** [pytorch/pytorch#21257]
## 🐛 Bug

Which distribution is `torch._multinomial_alias_draw` sampling from?
Related to #4115 and #18906.

## To Reproduce

Sampling 1 element at a time, hence, here we ignore whether it is sampling with or without replacement. In this case, the distribution to draw samples from equals the normalized weights. 

```python
import torch
import scipy.stats
import numpy as np
import torch.nn.functional as F

n = 10000
replace = True
device = 'cpu'

multinomial_alias_samples = []
multinomial_samples = []

weights = torch.Tensor([0.1, 0.6, 0.2, 0.1], device=device)
J, q = torch._multinomial_alias_setup(weights)

for _ in range(n):
	multinomial_alias_samples += torch._multinomial_alias_draw(
							      q, 
							      J, 
							      1
							    ).cpu().numpy().tolist()
	multinomial_samples += torch.multinomial(
						  weights,
						  1,
						  replace
						).cpu().numpy().tolist()

correct_dist =  weights / weights.sum()
correct_dist = correct_dist.to('cpu')
_, multinomial_alias_dist = np.unique(multinomial_alias_samples, return_counts=True)

_, p = scipy.stats.chisquare(multinomial_alias_dist, correct_dist.numpy() * n)
print("[ALIAS] Chi-Squared Test p-value {:.3f}".format(p))
multinomial_alias_dist = torch.Tensor(multinomial_alias_dist) / n
print("[ALIAS] KL Divergence {:.3f}".format(
					    F.kl_div(
						multinomial_alias_dist.log(), 
						correct_dist, 
						reduction='sum')
					    )
      )

_, multinomial_dist = np.unique(multinomial_samples, return_counts=True)
_, p = scipy.stats.chisquare(multinomial_dist, correct_dist.numpy() * n)
print("[NO ALIAS] Chi-Squared Test p-value {:.3f}".format(p))
multinomial_dist = torch.Tensor(multinomial_dist) / n
print("[NO ALIAS] KL Divergence {:.3f}".format(
					    F.kl_div(
						multinomial_dist.log(), 
						correct_dist, 
						reduction='sum')
					    )
      )

```
Results found: 

- [ALIAS] p-values < 0.001 and kl_div > 0.3
- [NO ALIAS] p-values > 0.1 and kl_div < 0.01

## Expected behavior

Sample from a multinomial distribution.

## Environment

Today's nightly version of PyTorch.

**Comment by vishwakftw** [comment]
cc: @neerajprad

**Comment by ezyang** [comment]
@amartya18x do you remember? (I know you wrote this code a long time ago haha)

**Comment by vishwakftw** [comment]
I think I've figured out the bug.

**Comment by vishwakftw** [comment]
With the local changes, this is the output from the reproduction script:
```
[ALIAS] Chi-Squared Test p-value 0.378
[ALIAS] KL Divergence 0.000
[NO ALIAS] Chi-Squared Test p-value 0.171
[NO ALIAS] KL Divergence 0.000
```

Without the local changes, this is the output from the reproduction script:
```
[ALIAS] Chi-Squared Test p-value 0.000
[ALIAS] KL Divergence 0.374
[NO ALIAS] Chi-Squared Test p-value 0.658
[NO ALIAS] KL Divergence 0.000
```

**Comment by neerajprad** [comment]
@LeviViana - Thanks for such a clear bug report. @vishwakftw - Could you put up a PR, or point me to your branch?

**Comment by vishwakftw** [comment]
The bug fix is present here: https://github.com/vishwakftw/pytorch/commit/55ae5db05421eb9f7dc18858e0bd5937f7546a3f

**Comment by Hacky-DH** [comment]
Hi which pytorch version releases this fix? I still found same bug in pytorch 1.1.0

**Comment by ezyang** [comment]
Try 1.2.0

**Comment by amartya18x** [comment]
> @amartya18x do you remember? (I know you wrote this code a long time ago haha)

Sorry @ezyang for this incredibly late response. I didn't receive the email notification for the mention somehow. (And yeah it has been a very long time haha! I do not recall much about this piece of code)

**Comment by Hacky-DH** [comment]
> Try 1.2.0

Thanks. It works in 1.2.0

**Comment by xpai** [comment]
Why has `torch.multinomial_alias_draw` been removed now (pytorch v1.8.0)? Is there any other alternative APIs?

**Comment by ezyang** [comment]
They were removed because it was private API that wasn't used anywhere in the PyTorch implementations. I guess if people found it useful we should reinstate it as a proper public API, resurrecting the implementation removed from https://github.com/pytorch/pytorch/pull/50489 cc @ngimel @nikitaved  @xpai can you open a feature request for it?
